### ISSUE TITLE
local math alphabets reset incorrectly after nested math

### ISSUE THREAD

**Issue by davidcarlisle** [latex3/latex2e#1101]
## Brief outline of the bug
https://tex.stackexchange.com/questions/689560/many-instances-of-declaresymbolfont-leads-to-odd-behavior

This is more or less the same isse as #1028 but without an explicit change of math version.
After a nested math list, the local math alphabet assignments in the outer list are set incorrectly/

## Minimal example showing the bug

```latex
\RequirePackage{latexbug} 
\documentclass{article}

\DeclareSymbolFont{abc1}{OT1}{cmr}{m}{n}
\DeclareSymbolFont{abc2}{OT1}{cmr}{m}{n}
\DeclareSymbolFont{abc3}{OT1}{cmr}{m}{n}
\DeclareSymbolFont{abc4}{OT1}{cmr}{m}{n}
\DeclareSymbolFont{abc5}{OT1}{cmr}{m}{n}
\DeclareSymbolFont{abc6}{OT1}{cmr}{m}{n}
\DeclareSymbolFont{abc7}{OT1}{cmr}{m}{n}
\DeclareSymbolFont{abc8}{OT1}{cmr}{m}{n}
\DeclareSymbolFont{abc9}{OT1}{cmr}{m}{n}
\DeclareSymbolFont{abca}{OT1}{cmr}{m}{n}
\DeclareMathAlphabet\mathscr{OT1}{cmr}{m}{it}
\DeclareMathAlphabet\mathfrak{U}{euf}{m}{n}


\def\x#1{\showthe\textfont14 }
\begin{document}
% incorrect
\x0
$\mathscr{F}_x
\hbox{$M\x1$\x2}
 \mathfrak{p}\x3$
\x4

% correct
$\mathscr{F}_x M \mathfrak{p}$
\x5

\end{document}

```

## Log file (required) and possibly PDF file  

[dd840.log](https://github.com/latex3/latex2e/files/11866059/dd840.log)

**Comment by davidcarlisle** [comment]
I'm not sure how best to fix this. 

In the test, the nested math is just `\hbox{$x$}` so the most efficient thing to do would be to do nothing so that as far as math font allocations go, the nested math just acts as if it were in the outer formula, so `$x$` does nothing and if it did have `\mathxxx` it would add to the mathalphabet count of the outer formula.

The alternative is to arrange that it effectively does a `\mathversion` reset so the nested math has its own local math alphabet setup that then has to be restored at the end of the group. This has some cost overhead (especially as these things usually in `\mathchoice` so get done 4 times, but would for example mean that you could use different `\mathxx` in different array cells with each cell acting as a local math alphabet scope.

for bold symbols in #1028 / #1030 we reset the math version but there you are explicitly changing math version so the costs seem more reasonable.

I haven't actually got either strategy working as I'd like for the test case above but would be interested to know others thoughts on strategy.

**Comment by davidcarlisle** [comment]
related

https://tex.stackexchange.com/a/706491/1090

**Comment by mbertucci47** [comment]
Just wanted to bump this because it bit me again for a poster presentation at a conference :-) Only found the wrong alphabet used after printing. Here's an example that doesn't show anything new but shows how easy it is to induce the behavior and how it depends on what comes both before and after it.
```tex
\documentclass{beamer}

\usefonttheme[onlymath]{serif}

\usepackage{libertinus}
\usepackage{libertinust1math}
\RequirePackage[frak=euler,scr=boondox,cal=euler,bb=esstix]{mathalpha}
\RequirePackage{stmaryrd}

\begin{document}
\begin{frame}
$\mathcal{P}$

$\mathbb{P}$

$
\mathbb{V}(f_*\mathscr{F}(d))^{T\text{-}\mathrm{everywhere}}
\biggl(\mathscr{F}\mathbb{V}\biggr)
$
\end{frame}
\end{document}
```
![Screenshot 2024-11-18 112006](https://github.com/user-attachments/assets/625d431e-a5cd-4c56-9ddd-4838d2658dc9)

The `\mathrm` shows up as `\mathscr` in the superscript. Obviously beamer and all the packages are setting up more math fonts than allowed, but I would call it a pretty normal setup for someone who wants to use fonts beside the defaults.

Here's a list of things that make the issue go away in this example.
- changing `beamer` to `article`
- deleting the `$\mathcal{P}$` line
- removing the second `\mathscr{F}` but not the first
- removing the `\text{-}`
- removing just the `\biggr` and `\biggl`
- getting rid of any of the packages or removing any one of the `mathalpha` options

If the `\mathbb{P}` and first `\mathbb{V}` are removed, the output is still wrong but different.
```tex
\documentclass{beamer}

\usefonttheme[onlymath]{serif}

\usepackage{libertinus}
\usepackage{libertinust1math}
\RequirePackage[frak=euler,scr=boondox,cal=euler,bb=esstix]{mathalpha}
\RequirePackage{stmaryrd}

\begin{document}
\begin{frame}
$\mathcal{P}$

$
(f_*\mathscr{F}(d))^{T\text{-}\mathrm{everywhere}}
\biggl(\mathscr{F}\mathbb{V}\biggr)
$
\end{frame}
\end{document}
```
![ss2](https://github.com/user-attachments/assets/b63dd808-078c-41ba-b16c-d6b43b9cb50c)

**Comment by davidcarlisle** [comment]
@mbertucci47  thanks for the example, just to confirm, adding

```
\setcounter{localmathalphabets}{0}
```

Fixes that case.  

This obviously missed the November LaTeX release but hopefully we should have something for a dev release soon.

**Comment by mbertucci47** [comment]
@davidcarlisle That's true but it's not hard to come up with examples where that doesn't work.
```tex
\documentclass{beamer}

\usefonttheme[onlymath]{serif}

\usepackage{libertinus}
\usepackage{libertinust1math}
\RequirePackage[frak=euler,scr=boondox,cal=euler,bb=esstix]{mathalpha}
\usepackage{stmaryrd}

\setcounter{localmathalphabets}{0}

\begin{document}
\begin{frame}
\[ \mathcal{T}_x \subseteq \mathscr{F}_x\otimes_{\mathcal{O}_{X,x}}\mathcal{O}_{X,x}/\mathfrak{m}_x^{r+1}:=\mathscr{F}.\]
\[
    \frac{[\mathbb{V}(f_*\mathscr{F}(d))^{T\text{-}\,\mathrm{everywhere}}]}{[\mathbb{V}(f_*\mathscr{F}(d))]}
    = \prod_{x\in X/S} \biggl(1-\frac{[T^c]_x}{[\mathbb{V}(\mathscr{P}^r_{\!/S}\mathscr{F}(d))]_x}t\biggr) + O(\mathbb{L}^{-\epsilon d})
\]
\end{frame}
\end{document}
```

**Comment by davidcarlisle** [comment]
Oh sure but the second case is the too many alphabets error that the localmathalphabet system is designed to avoid, so if you disable that by setting the counter to 0, then you get the traditional error if you use too many alphabets.
